AsyncHttpConsumer with docs for long polling/SSE

matthiask opened the pull request:
Refs #841, #524 

This is a very rough initial draft for an asynchronous HTTP consumer. 

I was torn a bit whether I should even add some code. For me, the documentation part is much more interesting and important for now. There certainly are ways to make the functionality more developer-friendly later, so maybe we do not want to decide what the recommended API should be yet.

Also, no synchronous version yet because going completely async is in some horrible way easier to understand for me  😊

TODO:

- [x] Tests
- [x] Default content type / charset

andrewgodwin commented:
Yup, that seems pretty reasonable. 

Things I think could be done:
* Accept both dicts and list of tuples for headers (don't see a reason why not here)
* Don't use AsgiRequest at all for now, sessions and auth should already be handled by the ASGI versions of their middleware anyway
* Docstrings everywhere

I don't think a sync version is needed, because if you want that, well, you have Django views!

matthiask commented:
I've added tests and docs and squashed the history into a single commit (no need for fixup commits fixing previous fixups). Some feedback would be awesome if you find the time. Thanks :-)

andrewgodwin commented on channels/generic/http.py:
```diff
@@ -0,0 +1,82 @@
+        The default behavior closes the response, and further messages on
+        the channel will be ignored.
+        """
+        assert isinstance(body, six.binary_type), "Body is not bytes"
```
No need to use `six` - Channels 3 is Python 3 only!

andrewgodwin commented on docs/topics/consumers.rst:
```diff
@@ -277,3 +281,66 @@ AsyncJsonWebsocketConsumer
+            await self.send_body(json.dumps(event).encode("utf-8"))
+
+Of course you can also use those primitives to implement a HTTP endpoint for
+`Server-sent events`_::
```
I'd prefer this as an inline link (I know, I'm weird, but the rest of the docs work like that and I like consistency)

matthiask commented:
👍 Thanks! 

I was thrown off by other uses of `six` in Channels.

andrewgodwin commented:
Alright, merged, thanks for this!

I didn't realise I had left some `six` in there, that's be missing them during the 1-to-2 conversion. I will go kill those now.

erm commented:
I found this in the docs https://channels.readthedocs.io/en/latest/topics/consumers.html#asynchttpconsumer, but it hasn't been included in a new release/made available on PyPI as of yet - how are new releases handled currently?

andrewgodwin commented:
They happen when I have the time - this next release is a major one, and work has been very busy recently, so it might be another week or two. You can use the master branch if you want to play with it, though - like Django core, master is usually quite stable.

alega19 commented:
@andrewgodwin 
> I don't think a sync version is needed, because if you want that, well, you have Django views!

Why? How to implement Long polling with views? It needs a consumer. AsyncHttpConsumer is good until no need to call sync functions (orm, etc..).
